packages/babel-core/src/errors/rewrite-stack-trace.ts

```diff
@@ -0,0 +1,171 @@
+ * were directly thrown in the config file.
+ */
+
+const ErrorToString = Function.call.bind(Error.prototype.toString);
+
+const SUPPORTED = !!Error.captureStackTrace;
+
+// We add some extra frames to Error.stackTraceLimit, so that we can respect
+// the original Error.stackTraceLimit even after removing all our internal
+// frames.
+// STACK_TRACE_LIMIT_DELTA should be bigger than the expected number of internal
+// frames, but not too big because capturing the stack trace is slow (this is
+// why Error.stackTraceLimit does not default to Infinity!).
+// Increase it if needed.
+const STACK_TRACE_LIMIT_DELTA = 100;
+
+const START_HIDNG = "startHiding - secret - don't use this - v1";
+const STOP_HIDNG = "stopHiding - secret - don't use this - v1";
+
+type CallSite = Parameters<typeof Error.prepareStackTrace>[1][number];
+
+const expectedErrors = new WeakSet<Error>();
+const virtualFrames = new WeakMap<Error, CallSite[]>();
+
+function CallSite(filename: string): CallSite {
+  // We need to use a prototype otherwise it breaks source-map-support's internals
+  return Object.create({
+    isNative: () => false,
+    isConstructor: () => false,
+    isToplevel: () => true,
+    getFileName: () => filename,
+    getLineNumber: () => undefined,
+    getColumnNumber: () => undefined,
+    getFunctionName: () => undefined,
+    getMethodName: () => undefined,
+    getTypeName: () => undefined,
+    toString: () => filename,
+  } as CallSite);
+}
+
+export function injcectVirtualStackFrame(error: Error, filename: string) {
+  if (!SUPPORTED) return;
+
+  let frames = virtualFrames.get(error);
+  if (!frames) virtualFrames.set(error, (frames = []));
+  frames.push(CallSite(filename));
+
+  return error;
+}
+
+export function expectedError(error: Error) {
+  if (!SUPPORTED) return;
+  expectedErrors.add(error);
+  return error;
+}
+
+export function beginHiddenCallStack<A extends unknown[], R>(
+  fn: (...args: A) => R,
+) {
+  if (!SUPPORTED) return fn;
+
+  return Object.defineProperty(
+    function (...args: A) {
+      setupPrepareStackTrace();
+      return fn(...args);
+    },
+    "name",
+    { value: STOP_HIDNG },
+  );
+}
+
+export function endHiddenCallStack<A extends unknown[], R>(
+  fn: (...args: A) => R,
+) {
+  if (!SUPPORTED) return fn;
+
+  return Object.defineProperty(
+    function (...args: A) {
+      return fn(...args);
+    },
+    "name",
+    { value: START_HIDNG },
+  );
+}
+
+function setupPrepareStackTrace() {
+  // This function is a singleton
+  // @ts-expect-error
+  // eslint-disable-next-line no-func-assign
+  setupPrepareStackTrace = () => {};
+
+  const { prepareStackTrace = defaultPrepareStackTrace } = Error;
+
+  Error.stackTraceLimit += STACK_TRACE_LIMIT_DELTA;
+
+  Error.prepareStackTrace = function stackTraceRewriter(err, trace) {
+    let newTrace = [];
+
+    const isExpected = expectedErrors.has(err);
+    let status = isExpected ? "hiding" : "unknown";
```

Comment:
Nit: Can you restrict the typing here? It seems to me `status` can only be `"hiding" | "showing" | "unknown"`.